combinetf2/workspace.py

```diff
@@ -0,0 +1,493 @@
+import os
+
+import h5py
+import hist
+import numpy as np
+import tensorflow as tf
+
+from wums import ioutils  # isort: skip
+
+axis_downUpVar = hist.axis.Regular(
+    2, -2.0, 2.0, underflow=False, overflow=False, name="downUpVar"
+)
+
+
+def getImpactsAxes(indata):
+    impact_names = list(indata.signals.astype(str)) + list(indata.systs.astype(str))
+    return hist.axis.StrCategory(impact_names, name="impacts")
+
+
+def getGlobalImpactsAxes(indata):
+    impact_names = list(indata.systs.astype(str)[indata.nsystnoconstraint :])
+    return hist.axis.StrCategory(impact_names, name="impacts")
+
+
+def getImpactsAxesGrouped(indata, bin_by_bin_stat=False):
+    impact_names_grouped = list(indata.systgroups.astype(str))
+    # impact data stat
+    impact_names_grouped.append("stat")
+    if bin_by_bin_stat:
+        # impact bin-by-bin stat
+        impact_names_grouped.append("binByBinStat")
+    return hist.axis.StrCategory(impact_names_grouped, name="impacts")
+
+
+def get_name_label_expected_hists(
+    name=None, label=None, prefit=False, variations=False, process_axis=None
+):
+    if name is None:
+        name = "hist"
+        name += "_prefit" if prefit else "_postfit"
+        if process_axis is None:
+            name += "_inclusive"
+        if variations:
+            name += "_variations"
+
+    if label is None:
+        label = "expected number of events, "
+        label = f"prefit {label}" if prefit else f"postfit {label}"
+        if process_axis is None:
+            label += "for all processes combined, "
+        if variations:
+            label += "with variations, "
+
+    return name, label
+
+
+class Workspace:
+    def __init__(self, outdir, outname, fitter, postfix=None, projections=[]):
+
+        channels = set(p[0] for p in projections)
+        self.projections = {c: [] for c in channels}
+        for p in projections:
+            self.projections[p[0]].append(p[1:])
+
+        self.results = {}
+        self.reset_results(fitter.indata.channel_info.keys())
+
+        # some information for the impact histograms
+        self.global_impact_axis = getGlobalImpactsAxes(fitter.indata)
+        self.impact_axis = getImpactsAxes(fitter.indata)
+        self.grouped_impact_axis = getImpactsAxesGrouped(
+            fitter.indata, fitter.binByBinStat
+        )
+
+        self.parms = fitter.parms
+        self.npoi = fitter.npoi
+        self.noigroupidxs = fitter.indata.noigroupidxs
+
+        self.extension = "hdf5"
+        self.file_path = self.get_file_path(outdir, outname, postfix)
+
+    def __enter__(self):
+        """Open the file when entering the context."""
+        print(f"Write results in file {self.file_path}")
+        self.fout = h5py.File(self.file_path, "w")
```

Comment:
I don't think this is the correct way of "wrapping" the context manager functionality (trying to find an appropriate example of the right way to do it though)

Reply:
so to do this consistently I think the self.fout =  statement should be moved to the end of __init__ and the __exit__ function should return False in order to properly re-raise any exceptions

Reply:
OK, moved the file opening to the __init__ function. I think where we open the file only matters in practice if the Workspace object is used not as context manager. I.e. as ```ws = Workspace()```. 
From https://book.pythontips.com/en/latest/context_managers.html I read that if nothing is returned in the __exit__ function the exception is raised, so it doesn't matter if we return False or nothing.

Reply:
Ok yes this is fine then.